### ISSUE TITLE
SSL Nagios

### ISSUE THREAD

**Issue by micheleberardi** [matteocorti/check_ssl_cert#48]
Hi
thanks for your script
i have a question 
i have a server web with multi-domains and virtual host on apache server
i have 2 domains
https://www.domain.com
https://www.domain2.com
when i run the check on https://www.domain2.com the check get the date of the SSL certificate of the first virtual host 
below an example
./check_ssl_cert -H www.domain2.com
i got an answer like that 
SSL_CERT OK - X.509 certificate '*.domain.com' from 'COMODO RSA Domain

any ideas ?

Thanks

**Comment by matteocorti** [comment]
Would it be possible to get the real URL so that I can test?

**Comment by micheleberardi** [comment]
sure
domain 1 tag.gurunetcdn.com
domain 2 tag.hera.pe

thanks for your help

**Comment by matteocorti** [comment]
Which version are you using? Which OpenSSL version?

I get

```
$ ./check_ssl_cert -H tag.hera.pe
SSL_CERT OK - X.509 certificate '*.hera.pe' from 'COMODO RSA Domain Validation Secure Server CA' valid until May 17 23:59:59 2019 GMT (expires in 735 days)|days=735;;;;
```

and

```
$ ./check_ssl_cert -H tag.gurunetcdn.com 
SSL_CERT OK - X.509 certificate '*.gurunetcdn.com' from 'COMODO RSA Domain Validation Secure Server CA' valid until Feb  8 23:59:59 2019 GMT (expires in 637 days)|days=637;;;;
```

You can also execute the script with the `-d` option and post the debugging output.

**Comment by micheleberardi** [comment]
Hi

thanks for the reply to me so quickly

my version of OpenSSL is

 OpenSSL version: OpenSSL 0.9.8e-fips-rhel5 01 Jul 2008

i tried the command with -d

this below the output

[./check_ssl_cert -H tag.hera.pe -d
[DBG] ROOT_CA =
expect not available
timeout not available
disabling timeouts
[DBG] perl available: /usr/bin/perl
[DBG] date available: /bin/date
found GNU date with timestamp support: enabling date computations
[DBG] check_ssl_version: 1.45.0
[DBG] OpenSSL binary: /usr/bin/openssl
[DBG] OpenSSL version: OpenSSL 0.9.8e-fips-rhel5 01 Jul 2008
[DBG] System info: Linux acous03nagios18.acoteldc.com 2.6.18-406.el5 #1 SMP
Tue Jun 2 17:25:57 EDT 2015 x86_64 x86_64 x86_64 GNU/Linux
[DBG] Date computation: GNU
'/usr/bin/openssl s_client' does not support '-servername': disabling
virtual server support
downloading certificate to /tmp
[DBG] executing with timeout (15s): echo 'Q' | /usr/bin/openssl s_client
-connect tag.hera.pe:443  -verify 6     2> /tmp/check_ssl_certF30349 1>
/tmp/check_ssl_certH30347
[DBG] storing a copy of the retrieved certificate in tag.hera.pe.crt
[DBG] storing a copy of the OpenSSL errors in tag.hera.pe.error
parsing the certificate file
unknown option -ocsp_uri
usage: x509 args
 -inform arg     - input format - default PEM (one of DER, NET or PEM)
 -outform arg    - output format - default PEM (one of DER, NET or PEM)
 -keyform arg    - private key format - default PEM
 -CAform arg     - CA format - default PEM
 -CAkeyform arg  - CA key format - default PEM
 -in arg         - input file - default stdin
 -out arg        - output file - default stdout
 -passin arg     - private key password source
 -serial         - print serial number value
 -subject_hash   - print subject hash value
 -issuer_hash    - print issuer hash value
 -hash           - synonym for -subject_hash
 -subject        - print subject DN
 -issuer         - print issuer DN
 -email          - print email address(es)
 -startdate      - notBefore field
 -enddate        - notAfter field
 -purpose        - print out certificate purposes
 -dates          - both Before and After dates
 -modulus        - print the RSA key modulus
 -pubkey         - output the public key
 -fingerprint    - print the certificate fingerprint
 -alias          - output certificate alias
 -noout          - no certificate output
 -ocspid         - print OCSP hash values for the subject name and public
key
 -trustout       - output a "trusted" certificate
 -clrtrust       - clear all trusted purposes
 -clrreject      - clear all rejected purposes
 -addtrust arg   - trust certificate for a given purpose
 -addreject arg  - reject certificate for a given purpose
 -setalias arg   - set certificate alias
 -days arg       - How long till expiry of a signed certificate - def 30
days
 -checkend arg   - check whether the cert expires in the next arg seconds
                   exit 1 if so, 0 if not
 -signkey arg    - self sign cert with arg
 -x509toreq      - output a certification request object
 -req            - input is a certificate request, sign and output.
 -CA arg         - set the CA certificate, must be PEM format.
 -CAkey arg      - set the CA key, must be PEM format
                   missing, it is assumed to be in the CA file.
 -CAcreateserial - create serial number file if it does not exist
 -CAserial arg   - serial file
 -set_serial     - serial number to use
 -text           - print the certificate in text form
 -C              - print out C code forms
 -<dgst>         - digest to use, see openssl dgst -h output for list
 -extfile        - configuration file with X509V3 extensions to add
 -extensions     - section from config file with X509V3 extensions to add
 -clrext         - delete extensions before signing and input certificate
 -nameopt arg    - various certificate name options
 -engine e       - use engine e, possibly a hardware device.
 -certopt arg    - various certificate text options
[DBG]  subject= OU = Domain Control Validated, OU = PositiveSSL Wildcard,
CN = *.gurunetcdn.com
[DBG] CN         = *.gurunetcdn.com
[DBG] CA_O       = COMODO CA Limited
[DBG] CA_CN      = COMODO RSA Domain Validation Secure Server CA
[DBG] SERIAL     = D13E39426407E688817102843E28E2A2
[DBG] OCSP_URI   =
[DBG] ISSUER_URI = http://crt.comodoca.com/COMODORSADomainValidationSecur
eServerCA.crt
[DBG]         Signature Algorithm: sha256WithRSAEncryption
[DBG] Date computations: GNU
The certificate will expire in 637 day(s)
[DBG] Checking expiration date
[DBG] OCSP: fetching issuer certificate http://crt.comodoca.com/
COMODORSADomainValidationSecureServerCA.crt to /tmp/check_ssl_certW30351
[DBG] OCSP: issuer certificate type:  data
[DBG] OCSP: converting issuer certificate from DER to PEM
[DBG] OCSP: storing a copy of the retrieved issuer certificate to
COMODORSADomainValidationSecureServerCA.crt
[DBG] OCSP: host =
openssl ocsp does not support the -header option: disabling OCSP checks
SSL_CERT OK - X.509 certificate '*.gurunetcdn.com' from 'COMODO RSA Domain
Validation Secure Server CA' valid until Feb  8 23:59:59 2019 GMT (expires
in 637 days)|days=637;;;;

with the new version of openssl it works

Thanks

**Comment by matteocorti** [comment]
I will try to test but OpenSSL 0.9.8 is no more supported and since RHEL is also no more supported the back porting is also no more available.

**Comment by matteocorti** [comment]
Your server uses SNI (https://en.wikipedia.org/wiki/Server_Name_Indication) while your version of OpenSSL does not support it:

> In 2004, a patch for adding TLS/SNI into OpenSSL was created by the EdelKey project. In 2006, this patch was then ported to the development branch of OpenSSL, and in 2007 it was back-ported to OpenSSL 0.9.8 (first released in 0.9.8f).

**Comment by micheleberardi** [comment]
Hi Matteo

now with the verion of openssl it works

./check_ssl_cert -H tag.hera.pe -c
SSL_CERT UNKNOWN tag.hera.pe: -c,--critical requires an argument
[root@localhost check_ssl_cert]# ./check_ssl_cert -H tag.hera.pe -c 30
SSL_CERT OK - X.509 certificate '*.hera.pe' from 'COMODO RSA Domain
Validation Secure Server CA' valid until May 17 23:59:59 2019 GMT (expires
in 730 days)|days=730;;30;;
[root@localhost check_ssl_cert]# ./check_ssl_cert -H tag.hera.pe -d
[DBG] ROOT_CA =
expect not available
timeout available (/usr/bin/timeout)
[DBG] perl available: /usr/bin/perl
[DBG] date available: /usr/bin/date
found GNU date with timestamp support: enabling date computations
[DBG] check_ssl_version: 1.45.0
[DBG] OpenSSL binary: /usr/bin/openssl
[DBG] OpenSSL version: OpenSSL 1.0.1e-fips 11 Feb 2013
[DBG] System info: Linux localhost.localdomain 3.10.0-514.16.1.el7.x86_64
#1 SMP Wed Apr 12 15:04:24 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
[DBG] Date computation: GNU
[DBG] '/usr/bin/openssl s_client' supports '-servername': using -servername
tag.hera.pe
downloading certificate to /tmp
[DBG] executing with timeout (15s): echo 'Q' | /usr/bin/openssl s_client
-connect tag.hera.pe:443 -servername tag.hera.pe -verify 6     2>
/tmp/check_ssl_certR97nuS 1> /tmp/check_ssl_cert948WZ1
[DBG]   /usr/bin/timeout 15 /bin/sh -c "echo 'Q' | /usr/bin/openssl
s_client   -connect tag.hera.pe:443 -servername tag.hera.pe -verify 6
2> /tmp/check_ssl_certR97nuS 1> /tmp/check_ssl_cert948WZ1"
[DBG] storing a copy of the retrieved certificate in tag.hera.pe.crt
[DBG] storing a copy of the OpenSSL errors in tag.hera.pe.error
parsing the certificate file
[DBG]  subject= OU = Domain Control Validated, OU = EssentialSSL Wildcard,
CN = *.hera.pe
[DBG] CN         = *.hera.pe
[DBG] CA_O       = COMODO CA Limited
[DBG] CA_CN      = COMODO RSA Domain Validation Secure Server CA
[DBG] SERIAL     = 61CF209497EBD4FF6195A57A5DB68B2A
[DBG] OCSP_URI   = http://ocsp.comodoca.com
[DBG] ISSUER_URI =
http://crt.comodoca.com/COMODORSADomainValidationSecureServerCA.crt
[DBG]     Signature Algorithm: sha256WithRSAEncryption
[DBG] Date computations: GNU
The certificate will expire in 730 day(s)
[DBG] Checking expiration date
[DBG] OCSP: fetching issuer certificate
http://crt.comodoca.com/COMODORSADomainValidationSecureServerCA.crt to
/tmp/check_ssl_cert0zlRd1
[DBG] OCSP: issuer certificate type:  data
[DBG] OCSP: converting issuer certificate from DER to PEM
[DBG] OCSP: storing a copy of the retrieved issuer certificate to
COMODORSADomainValidationSecureServerCA.crt
[DBG] OCSP: host = ocsp.comodoca.com
[DBG] openssl ocsp support the -header option
[DBG] openssl ocsp -header requires 'key value'
[DBG] executing /usr/bin/openssl ocsp -no_nonce -issuer
/tmp/check_ssl_cert0zlRd1 -cert /tmp/check_ssl_cert948WZ1  -url
http://ocsp.comodoca.com  -header HOST ocsp.comodoca.com
[DBG] OCSP: response = Response Verify Failure
[DBG] OCSP: response = 140179206981536:error:27069076:OCSP
routines:OCSP_basic_verify:signer certificate not found:ocsp_vfy.c:85:
[DBG] OCSP: response = /tmp/check_ssl_cert948WZ1: good
[DBG] OCSP: response = This Update: May 16 18:45:56 2017 GMT
[DBG] OCSP: response = Next Update: May 23 18:45:56 2017 GMT
SSL_CERT OK - X.509 certificate '*.hera.pe' from 'COMODO RSA Domain
Validation Secure Server CA' valid until May 17 23:59:59 2019 GMT (expires
in 730 days)|days=730;;;;


Thanks
